lib/src/signed_video_h26x_sign.c

```diff
@@ -65,6 +73,92 @@ h26x_set_nal_uuid_type(signed_video_t *self, uint8_t **payload, SignedVideoUUIDT
   }
 }
 
+/* Frees all payloads in the |payload_buffer|. Declared in signed_video_internal.h */
+void
+free_payload_buffer(uint8_t *payload_buffer[])
+{
+  for (int i = 0; i < MAX_NALUS_TO_PREPEND; i++) {
+    // Note that the first location of the payload pointer pair points to the location of the
+    // memory.
+    free(payload_buffer[2 * i]);
+    payload_buffer[2 * i] = NULL;
+    payload_buffer[2 * i + 1] = NULL;
+  }
+}
+
+/* Adds the |payload| to the next available slot in |payload_buffer|. */
+static void
+add_payload_to_buffer(signed_video_t *self, uint8_t *payload, uint8_t *payload_signature_ptr)
+{
+  assert(self);
+
+  if (self->payload_buffer_idx >= MAX_NALUS_TO_PREPEND) {
+    // Not enough space for this payload. Free the memory and return.
+    free(payload);
+    return;
+  }
+
+  self->payload_buffer[2 * self->payload_buffer_idx] = payload;
+  self->payload_buffer[2 * self->payload_buffer_idx + 1] = payload_signature_ptr;
+  self->payload_buffer_idx += 1;
+}
+
+/* Picks the oldest payload from the payload_buffer and completes it with the generated signature.
+ * If we have no signature the SEI payload is freed and not added to the video session. */
+static svi_rc
+complete_sei_nalu_and_add_to_prepend(signed_video_t *self)
+{
+  assert(self);
+
+  // Get the oldest payload.
+  uint8_t *payload = self->payload_buffer[0];
+  uint8_t *payload_signature_ptr = self->payload_buffer[1];
+  SignedVideoPrependInstruction prepend_instruction = SIGNED_VIDEO_PREPEND_NOTHING;
+  size_t data_size = 0;
+  svi_rc status = SVI_UNKNOWN;
+
+  // If the signature could not be generated |signature_size| equals zero. Free the started SEI and
+  // move on. This is a valid operation. What will happen is that the video will have an unsigned
+  // GOP.
+  if (self->signature_info->signature_size == 0) {
+    signed_video_nalu_data_free(payload);
+    payload = NULL;
+    status = SVI_OK;
+    goto done;
+  } else if (!payload) {
+    // No more pending payloads. Already freed due to too many unsigned SEIs.
+    status = SVI_OK;
+    goto done;
+  }
+
+  SVI_TRY()
+    // Add the signature to the SEI payload.
+    data_size = get_sign_and_complete_sei_nalu(self, &payload, payload_signature_ptr);
+    SVI_THROW_IF(!data_size, SVI_UNKNOWN);
```

Comment:
Maybe have "get_sign_and_complete_sei_nalu" inside SVI_THROW_IF?

Reply:
Unfortunately, `data_size` is needed on line 143

Reply:
Good.